File: sdk/go/namespaces.go
package bsdkrun

import (
	"encoding/json"
	"strings"
)

// Host-level operations that aren't tied to a single machine, grouped into
// the same namespaces the Python SDK ships as submodules:
// bsdkrun.Images.List(), bsdkrun.Volumes.Remove(...),
// bsdkrun.Networks.Create(...), bsdkrun.System.Probe().

// ImagesNamespace groups host-level image operations; use the Images var.
type ImagesNamespace struct{}

// Images is the image namespace.
var Images ImagesNamespace

// List lists downloaded images (pulled OCI images + fetched BSD images).
func (ImagesNamespace) List() ([]ImageInfo, error) {
	res, err := RunChecked([]string{"images", "--json"}, "bsdkrun images", nil)
	if err != nil {
		return nil, err
	}
	var rows []ImageInfo
	if err := json.Unmarshal([]byte(orEmptyJSON(res.Stdout)), &rows); err != nil {
		return nil, err
	}
	return rows, nil
}

// VolumesNamespace groups host-level volume operations; use the Volumes var.
type VolumesNamespace struct{}

// Volumes is the volume namespace.
var Volumes VolumesNamespace

// List lists persistent volumes.
func (VolumesNamespace) List() ([]VolumeInfo, error) {
	res, err := RunChecked([]string{"volume", "ls", "--json"}, "bsdkrun volume ls", nil)
	if err != nil {
		return nil, err
	}
	var rows []VolumeInfo
	if err := json.Unmarshal([]byte(orEmptyJSON(res.Stdout)), &rows); err != nil {
		return nil, err
	}
	return rows, nil
}

// Remove removes one or more volumes (and their data).
func (VolumesNamespace) Remove(names ...string) error {
	return removeVolumes(false, names)
}

// ForceRemove removes one or more volumes even when in use.
func (VolumesNamespace) ForceRemove(names ...string) error {
	return removeVolumes(true, names)
}

func removeVolumes(force bool, names []string) error {
	args := []string{"volume", "rm"}
	if force {
		args = append(args, "--force")
	}
	args = append(args, names...)
	_, err := RunChecked(args, "bsdkrun volume rm", nil)
	return err
}

// NetworksNamespace groups global-network operations — reach machines by
// name on a shared subnet; use the Networks var.
type NetworksNamespace struct{}

// Networks is the network namespace.
var Networks NetworksNamespace

// List lists global networks and their member counts.
func (NetworksNamespace) List() ([]NetworkInfo, error) {
	res, err := RunChecked([]string{"network", "ls", "--json"}, "bsdkrun network ls", nil)
	if err != nil {
		return nil, err
	}
	var rows []NetworkInfo
	if err := json.Unmarshal([]byte(orEmptyJSON(res.Stdout)), &rows); err != nil {
		return nil, err
	}
	return rows, nil
}

// Create creates a global network (starts its shared switch).
func (NetworksNamespace) Create(name string) error {
	_, err := RunChecked([]string{"network", "create", name}, "bsdkrun network create", nil)
	return err
}

// Remove removes one or more networks.
func (NetworksNamespace) Remove(names ...string) error {
	return removeNetworks(false, names)
}

// ForceRemove removes one or more networks even with live members.
func (NetworksNamespace) ForceRemove(names ...string) error {
	return removeNetworks(true, names)
}

func removeNetworks(force bool, names []string) error {
	args := []string{"network", "rm"}
	if force {
		args = append(args, "--force")
	}
	args = append(args, names...)
	_, err := RunChecked(args, "bsdkrun network rm", nil)
	return err
}

// Connect joins or switches a machine (by id or name) to a network (next
// start).
func (NetworksNamespace) Connect(machine, network string) error {
	_, err := RunChecked([]string{"network", "connect", machine, network}, "bsdkrun network connect", nil)
	return err
}

// Disconnect detaches a machine from its network. Applies on its next
// start.
func (NetworksNamespace) Disconnect(machine string) error {
	_, err := RunChecked([]string{"network", "disconnect", machine}, "bsdkrun network disconnect", nil)
	return err
}

// Sync refreshes members' /etc/hosts so peers resolve by name (notably
// NetBSD).
func (NetworksNamespace) Sync(network string) error {
	_, err := RunChecked([]string{"network", "sync", network}, "bsdkrun network sync", nil)
	return err
}

// Members lists the machines currently attached to network (running or
// stopped).
func (NetworksNamespace) Members(network string) ([]SandboxInfo, error) {
	rows, err := ListSandboxes(true)
	if err != nil {
		return nil, err
	}
	var out []SandboxInfo
	for _, info := range rows {
		if info.Network == network {
			out = append(out, info)
		}
	}
	return out, nil
}

// SystemNamespace groups host toolchain / image operations; use the System
// var.
type SystemNamespace struct{}

// System is the system namespace.
var System SystemNamespace

// Probe sanity-checks the toolchain (libkrun links, a context is
// creatable). It does not boot. Returns true on success.
func (SystemNamespace) Probe() bool {
	res, err := Run([]string{"probe"}, nil)
	return err == nil && res.ExitCode == 0
}

// FetchImageBuilder downloads + prepares a BSD image ahead of time.
type FetchImageBuilder struct {
	osName  string
	version string
	dir     string
	force   bool
}

// FetchImage starts a fetch for "freebsd" or "netbsd": chain
// Version/Dir/Force, then Run.
func (SystemNamespace) FetchImage(osName string) *FetchImageBuilder {
	return &FetchImageBuilder{osName: osName}
}

// Version picks the release to fetch.
func (b *FetchImageBuilder) Version(version string) *FetchImageBuilder {
	b.version = version
	return b
}

// Dir sets the download directory.
func (b *FetchImageBuilder) Dir(dir string) *FetchImageBuilder {
	b.dir = dir
	return b
}

// Force re-downloads even when cached.
func (b *FetchImageBuilder) Force() *FetchImageBuilder {
	b.force = true
	return b
}

// Run performs the fetch and returns the command output.
func (b *FetchImageBuilder) Run() (string, error) {
	args := []string{"fetch", "--os", b.osName}
	if b.version != "" {
		args = append(args, "--version", b.version)
	}
	if b.dir != "" {
		args = append(args, "--dir", b.dir)
	}
	if b.force {
		args = append(args, "--force")
	}
	res, err := RunChecked(args, "bsdkrun fetch", nil)
	if err != nil {
		return "", err
	}
	return res.Stdout, nil
}

// Versions lists the arm64 builds available to fetch for a BSD
// ("freebsd"/"netbsd") as the non-empty output lines.
func (SystemNamespace) Versions(osName string) ([]string, error) {
	res, err := RunChecked([]string{"versions", "--os", osName}, "bsdkrun versions", nil)
	if err != nil {
		return nil, err
	}
	var out []string
	for _, line := range strings.Split(res.Stdout, "\n") {
		if trimmed := strings.TrimSpace(line); trimmed != "" {
			out = append(out, trimmed)
		}
	}
	return out, nil
}

// GrowDisk grows a raw disk image (the guest expands its root FS on next
// boot).
func (SystemNamespace) GrowDisk(disk, size string) error {
	_, err := RunChecked([]string{"grow", "--disk", disk, "--size", size}, "bsdkrun grow", nil)
	return err
}

func orEmptyJSON(stdout string) string {
	if strings.TrimSpace(stdout) == "" {
		return "[]"
	}
	return stdout
}

File: sdk/scala/src/main/scala/bsdkrun/Namespaces.scala
package bsdkrun

import bsdkrun.Types.{ImageInfo, NetworkInfo, VolumeInfo}

/** Downloaded images — OCI images pulled and BSD images fetched. */
object Images:

  /** List downloaded images. */
  def list(): Either[BsdkrunError, Seq[ImageInfo]] =
    Proc
      .runChecked(Seq("images", "--json"), "bsdkrun images")
      .flatMap(res => Types.rows(res.stdout, "bsdkrun images", Types.imageInfo))

/** Persistent volumes. */
object Volumes:

  def list(): Either[BsdkrunError, Seq[VolumeInfo]] =
    Proc
      .runChecked(Seq("volume", "ls", "--json"), "bsdkrun volume ls")
      .flatMap(res => Types.rows(res.stdout, "bsdkrun volume ls", Types.volumeInfo))

  /** Remove volumes and their data. */
  def remove(names: Seq[String], force: Boolean = false): Either[BsdkrunError, Unit] =
    val args = Seq("volume", "rm") ++ (if force then Seq("--force") else Nil) ++ names
    Proc.runChecked(args, "bsdkrun volume rm").map(_ => ())

/** Global networks — a shared subnet plus internal DNS, so machines reach each
  * other by name.
  */
object Networks:

  def list(): Either[BsdkrunError, Seq[NetworkInfo]] =
    Proc
      .runChecked(Seq("network", "ls", "--json"), "bsdkrun network ls")
      .flatMap(res => Types.rows(res.stdout, "bsdkrun network ls", Types.networkInfo))

  def create(name: String): Either[BsdkrunError, Unit] =
    Proc.runChecked(Seq("network", "create", name), "bsdkrun network create").map(_ => ())

  def remove(names: Seq[String], force: Boolean = false): Either[BsdkrunError, Unit] =
    val args = Seq("network", "rm") ++ (if force then Seq("--force") else Nil) ++ names
    Proc.runChecked(args, "bsdkrun network rm").map(_ => ())

  /** Attach a machine to a network. Takes effect on its next start. */
  def connect(network: String, id: String): Either[BsdkrunError, Unit] =
    Proc.runChecked(Seq("network", "connect", network, id), "bsdkrun network connect").map(_ => ())

  def disconnect(id: String): Either[BsdkrunError, Unit] =
    Proc.runChecked(Seq("network", "disconnect", id), "bsdkrun network disconnect").map(_ => ())

  /** Re-apply a network's DNS and addressing to its running members. */
  def sync(name: String): Either[BsdkrunError, Unit] =
    Proc.runChecked(Seq("network", "sync", name), "bsdkrun network sync").map(_ => ())

/** Host-level toolchain and image operations.
  *
  * Named `Host` rather than `System` — which is what the other SDKs call it —
  * because `java.lang.System` is auto-imported into every Scala file. An
  * `object System` here would shadow it for anyone doing `import bsdkrun.*`,
  * so `System.currentTimeMillis()` would stop compiling in their code.
  */
object Host:

  /** Sanity-check the toolchain: verify libkrun links and a VM context can be
    * created and configured. Does not boot anything.
    */
  def probe(): Boolean = Proc.run(Seq("probe")).exists(_.exitCode == 0)

  /** Check that this host can run machines, and report what to fix if not.
    *
    * Returns the report as JSON — `ok` says whether anything failed, and
    * `checks` lists each one with a `fix` where there is something to do.
    */
  def doctor(): Either[BsdkrunError, ujson.Value] =
    Proc.run(Seq("doctor", "--json")).flatMap { res =>
      // `doctor` exits 1 when a check fails, which is the answer rather than a
      // failure to get one — so the report is parsed either way.
      Types.one(res.stdout, "bsdkrun doctor")
    }

  /** Download and prepare a BSD image ahead of time. */
  def fetchImage(
      os: String,
      version: Option[String] = None,
      force: Boolean = false
  ): Either[BsdkrunError, String] =
    val args = Seq("fetch", "--os", os) ++
      version.toSeq.flatMap(v => Seq("--version", v)) ++
      (if force then Seq("--force") else Nil)
    Proc.runChecked(args, "bsdkrun fetch").map(_.stdout)

  /** The builds available to fetch for a BSD, one per line. */
  def versions(os: String): Either[BsdkrunError, Seq[String]] =
    Proc
      .runChecked(Seq("versions", "--os", os), "bsdkrun versions")
      .map(_.stdout.linesIterator.map(_.trim).filter(_.nonEmpty).toSeq)

  /** Grow a raw disk image. Only ever enlarges the file. */
  def growDisk(disk: String, size: String): Either[BsdkrunError, Unit] =
    Proc.runChecked(Seq("grow", "--disk", disk, "--size", size), "bsdkrun grow").map(_ => ())
